Sources/SwiftSyntaxMacroExpansion/MacroExpansion.swift

```diff
@@ -0,0 +1,203 @@
+import SwiftSyntax
+import SwiftSyntaxMacros
+
+public enum MacroRole {
+  case expression
+  case declaration
+  case accessor
+  case memberAttribute
+  case member
+  case peer
+  case conformance
+  case codeItem
+}
+
+/// Simple diagnostic message
+private struct MacroExpansionError: Error, CustomStringConvertible {
+  let description: String
+}
+
+/// Expand `@freestanding(XXX)` macros.
+public func expandFreestandingMacro(
+  definition: Macro.Type,
+  node: FreestandingMacroExpansionSyntax,
+  in context: some MacroExpansionContext
+) -> String? {
+  do {
+    func _expand(node: some FreestandingMacroExpansionSyntax) throws -> String {
+      let expandedSyntax: Syntax
+      switch definition {
+      case let exprMacroDef as ExpressionMacro.Type:
+        expandedSyntax = try Syntax(exprMacroDef.expansion(of: node, in: context))
+
+      case let declMacroDef as DeclarationMacro.Type:
+        let rewritten = try declMacroDef.expansion(of: node, in: context)
+        expandedSyntax = Syntax(
+          CodeBlockItemListSyntax(
+            rewritten.map {
+              CodeBlockItemSyntax(item: .decl($0))
+            }
+          )
+        )
+
+      case let codeItemMacroDef as CodeItemMacro.Type:
+        let rewritten = try codeItemMacroDef.expansion(of: node, in: context)
+        expandedSyntax = Syntax(CodeBlockItemListSyntax(rewritten))
+
+      default:
+        throw MacroExpansionError(description: "macro doesn't conform to required macro role")
+      }
+      return expandedSyntax.formattedExpansion(definition.formatMode)
+    }
+    return try _openExistential(node, do: _expand)
+  } catch {
+    context.addDiagnostics(from: error, node: node)
+    return nil
+  }
+}
+
+/// Expand `@attached(XXX)` macros.
+public func expandAttachedMacro<Context: MacroExpansionContext>(
+  definition: Macro.Type,
+  macroRole: MacroRole,
+  attributeNode: AttributeSyntax,
+  declarationNode: DeclSyntax,
+  parentDeclNode: DeclSyntax?,
+  in context: Context
+) -> [String]? {
```

Comment:
Would it maybe be cleaner to make this method `throw` instead of returning an optional? I think requiring callers to add the error message to the context’s diagnostics isn’t too much to ask.

Reply:
I think there's no reason _not_ to catch exceptions, turn it to a diagnostic and emit it to the context in this function. The whole point of this PR is to minimize the duplicated logic. Why do you think making it `throws` is cleaner?

Reply:
I think I just have a disliking of making `nil` indicate the error case because it doesn’t allow the client to check what went wrong. Since this is `public` anyone could call it to try expanding a macro node and for their use case, it might be useful to know what went wrong, even though for our use cases it doesn’t matter.

That being said, I think I’m fine with the current approach if we add a comment saying that upon failure (aka. if this returns `nil`) a diagnostic is guaranteed to be added to `context`.